### ISSUE TITLE
Check migrations and apply them

### ISSUE THREAD

**Issue by AbelHristodor** [django-hurricane/django-hurricane#106]
It would be nice to have a command that checks the migrations (like --check-migrations) but then also applies them if there are any.

I've found that django apps, especially large ones, when scaling and/or deploying to k8s, since they all have to connect to db and maybe also refresh some tables (like pg-views does) they become very slow to start, even if there aren't even any migrations to run.

It would be nice to have a --check-migrations-apply command that checks migrations, and then applies them if there are any. I tried to make a pr but i guess i don't have enough permissions.

```python
def check_db_and_migrations_and_apply(webhook_url: str = None, loop: asyncio.unix_events.SelectorEventLoop = None):
    try:
        while check_databases():
            number_of_migrations = count_migrations()

            logger.info(f"There are {number_of_migrations} pending migrations")
            if number_of_migrations == 0:
                logger.info("No pending migrations")
                break

            logger.info("Applying migrations")
            call_command("migrate")

    except Exception as e:
        error_trace = traceback.format_exc()
        logger.info("Webhook with a status warning has been initiated")

        StartupWebhook().run(
            url=webhook_url or None, error_trace=error_trace, close_loop=True, status=WebhookStatus.WARNING, loop=loop
        )
        raise e
```

I'm more than happy to contribute.

**Comment by tschale** [comment]
@AbelHristodor Thank you for your contribution! You should be able to achieve that with the `--command` flag, i.e. `--command migrate`. Or is this not working for you?

Just a further comment for your interest. We have solved the migration step for some of our applications differently. It can be problematic when using a data migration and multiple pods spin up at more or less the same time. Until the data migration is finished, every further pod will itself run the data migration. If the data migration is not idempotent (which they sometimes can't be), it will be applied multiple times, causing undesired behavior.

**Comment by AbelHristodor** [comment]
Yes, I agree that with the `--command migrate` command you can achieve something similar but not equal. From my own testing I found out that even running `migrate` without any real migrations that need to be applied, the app take much longer to spin up.

I don't really know why, but I know that since I've added a `check_migrations` command in my django app (that does what I stated above), and then spin it up with `python manage.py check_migrations && python manage.py runserver`, pods are so much faster to boot up, like really fast. Maybe because I have some packages that refresh some tables with some information (e.g. pg_views) which take some time. Just imagine having 6-7 pods that try to refresh some tables all at the same time.

I know for sure that, for example, when doing a `rolling-update` in k8s, only the first pod applies the migrations and then the others skip everything and simply boot up.

Just to give some numbers, before my django app (which is a production-grade app, not extremely heavy but not lightweight either) was taking ~40s, if there were like 5 pods all starting at the same time they'd take almost ~110s. With this command they only take ~25s, regardless of how many replicas are there.

**Comment by AbelHristodor** [comment]
Also, I have a question about resources, is it normal that w/hurricane RAM usage is very very high? (I got 2x more than plain Django)

**Comment by tschale** [comment]
> Also, I have a question about resources, is it normal that w/hurricane RAM usage is very very high? (I got 2x more than plain Django)

We haven't observed that so far on our deployments. We didn't run an extensive benchmark though, but judging from different resource requests/limits, for us the resource consumption is similar to when we deployed Django applications with uwsgi.


> I know for sure that, for example, when doing a `rolling-update` in k8s, only the first pod applies the migrations and then the others skip everything and simply boot up.

In general this works for us too, but as I've said we were able to observe the described behavior on a data migration. This is not to say that our deployment/migration process was optimal, but to me it seems like the possibility is there.



> Just to give some numbers, before my django app (which is a production-grade app, not extremely heavy but not lightweight either) was taking ~40s, if there were like 5 pods all starting at the same time they'd take almost ~110s. With this command they only take ~25s, regardless of how many replicas are there.

It's interesting to hear that, thank you for the insights! So far this doesn't seem to be an issue with our applications, but obviously from your report that can occur. You should be able to create a fork of this repository and create a pull request from that. Can you please to this for an additional command `--check-migrations-apply`?

**Comment by tschale** [comment]
Nice, thank you! You're quick :)

I'll give it a review today or tomorrow.

**Comment by tschale** [comment]
I'm gonna leave this open for now, as we haven't deployed a new release yet. We currently have issues with our testing pipeline (#111), due to which we are currently blocked on making a new release.
